### ISSUE TITLE
Filtering properties that relates to another documents in $getMembers()

### ISSUE THREAD

**Issue by Yahaire** [CarbonLDP/carbonldp-js-sdk#346]
**TL;DR**
_It would be amazing to treat access points like properties with literals when used in `.filter()`. Their literal value could be their `$id`._
.........

I have a database that consists of a container with movie documents, and a container with person documents. Movies and Persons are related through access points (Persons are actors, authors, writers, etc. of Movies).

**I am trying to search for all of the movies where a certain actor has appeared.** For this, I use the following method that searches through the Movies container:

```
const actorOriginalId = 'https://www.imdb.com/name/nm0001273/' // Sample actor: Lillian Gish

const rawResults = await this.carbonLDP.documents.$getMembers(this.moviesURL, _ => _
      .properties({
        'originalId': _.inherit,
        'label': _.inherit,
        'actor': { // My access point to persons
          "query": _ => _
            .properties({
              '$id': _.inherit,
              'originalId': _.inherit,
            })
            .filter(_.property('originalId') + ' = ' + _.value(actorOriginalId)) // Show only Lillian
        }
      })
      .filter(_.property('originalId') + ' = ' + _.value(actorOriginalId)) // Discard movies where actor is not Lillian
      .orderBy('originalId', 'DESC')
    );
```
The way this method works is that it queries through all of the movies, getting `$id` and `originalId` of the given actor should he be present in the actor access point. If the actor is present, this returns it as a property that has `$id` and `originalId` of the current movie in the main query. The main query filters out all of the movies that don't have a property of `actor` with the given `originalId`. _(As a side note, would have preferred to use the `id`, but, for some reason, I could never get a match with that property in getMembers)_

**This works**, returning only the movies where the given actor participates. This said, **there are quite a few problems with this method**:
- The subquery is ran through all of the movies, making the request take much longer and consume more resources on the server side.
- The method seems too complex for what it wants to achieve.
- After getting the results, one must resolve them to get all properties (See [my comment](https://github.com/CarbonLDP/carbonldp-js-sdk/issues/197#issuecomment-451250583) on issue #197)
- It cannot be paginated (See issue #296)

**It would be pretty awesome if**, instead of this subquerying, **we could rely on the `$id` found on the shallow document that the access point gives access to**. This could result in a method similar to this:
```
const actorId = 'https://movie-demo-ldp.carbonldp.com/persons/nm0001273/'; // Again, Lillian Gish, but now represented by its CarbonLDP Id

const rawResults = await this.carbonLDP.documents.$getMembers(this.moviesURL, _ => _
    .properties({
        'originalId': _.inherit,
        'label': _.inherit,
        'actor': _.inherit // No need to get extra information on the authors, we only want the movies!
    })
    .filter(_.property('actor.$id') + ' = ' + _.value(actorId)) // Discard movies where actor doesn't contain Lillian
    .orderBy('originalId', 'DESC')
);
```
This approach works with properties that have literal values, but it seems to be no good for access points. I believe it could prove very powerful, though! It would make some searches much faster while keeping the user utilizing only the SDK; no SPARQL needed.

Anyways, if you've reached this point, thanks a lot for your time!

**Comment by roddolf** [comment]
Just as a clarification, remember that an Access Point is just a document that manages a certain property of another document. So it doesn't affect if the property `actor` is set manually or by an Access Point.

Returning to your questions, the behavior of the `filter` method is the expected, it works locally in the query you are setting, and since every property is optional by default it will only be applied when content in the property `actor` exists.

There are multiple forms to archive what you desire.

1. Set the property as required so query only the movies that have authors:
```typescript
const actorOriginalId = 'https://www.imdb.com/name/nm0001273/' // Sample actor: Lillian Gish

const rawResults = await this.carbonLDP.documents.$getMembers(this.moviesURL, _ => _
      .properties({
        "actor": {
          "required": true,
          "query": _ => _
            .properties({
              "originalId": _.inherit,
            })
            .filter(_.property("originalId") + ' = ' + _.value(actorOriginalId))
        }
      })
    );
```

2. Extract the property filter into the main query using the reference to a sub-property (`author.originalId`):
```typescript
const actorOriginalId = 'https://www.imdb.com/name/nm0001273/' // Sample actor: Lillian Gish

const rawResults = await this.carbonLDP.documents.$getMembers(this.moviesURL, _ => _
      .properties({
        "actor": {
          "query": _ => _
            .properties({
              "originalId": _.inherit,
            })
        }
      })
      .filter(_.property("author.originalId") + ' = ' + _.value(actorOriginalId))
    );
```

3. Using the URI of the document.
  As you pointed out, this would be the optimal one since no sub-query is required.
  But remember that the dollar sing properties are only properties set by the SDK, so `$id` doesn't really exists in the data and is not a property you can query, it is the representation of the URI that represents the document. So to match the URI you need to match the property (_.property("author")) with the URI you have, using the method `_.object(actorId)`:
```typescript
const rawResults = await this.carbonLDP.documents.$getMembers(this.moviesURL, _ => _
      .properties({
        "actor": _.inherit
      })
      .filter(_.property("author") + ' = ' + _.object("https://example.com/actors/wanted-actor/"))
    );
```
